core/templates/services/literally-canvas-helper.service.ts

```diff
@@ -58,31 +58,240 @@ angular.module('oppia').factory('LiterallyCanvasHelperService', [
+          endCapShapes: null,
+          order: null,
+          tailSize: null,
+          pointCoordinatePairs: null,
+          smoothedPointCoordinatePairs: null,
+          pointSize: null,
+          pointColor: null,
+          smooth: null,
+          isClosed: null,
+          text: null,
+          font: null,
+          forcedWidth: null,
+          forcedHeight: null
+        },
+        id: ''
+      };
+      if (id[0] === 'ellipse') {
+        shape.className = 'Ellipse';
+        shape.data.x = node.attributes.cx.value - node.attributes.rx.value;
+        shape.data.y = node.attributes.cy.value - node.attributes.ry.value;
+        shape.data.width = 2 * node.attributes.rx.value;
+        shape.data.height = 2 * node.attributes.ry.value;
+        shape.data.strokeWidth = parseInt(
+          node.attributes['stroke-width'].value);
+        shape.data.strokeColor = node.attributes.stroke.value;
+        shape.data.fillColor = node.attributes.fill.value;
+        shape.id = id.slice(1).join('-');
+      } else if (id[0] === 'rectangle') {
+        shape.className = 'Rectangle';
+        var strokeWidth = node.attributes['stroke-width'].value;
+        var shift = strokeWidth % 2 !== 0 ? 0.5 : 0;
+        shape.data.x = node.attributes.x.value - shift;
+        shape.data.y = node.attributes.y.value - shift;
+        shape.data.width = parseInt(node.attributes.width.value);
+        shape.data.height = parseInt(node.attributes.height.value);
+        shape.data.strokeWidth = parseInt(strokeWidth);
+        shape.data.strokeColor = node.attributes.stroke.value;
+        shape.data.fillColor = node.attributes.fill.value;
+        shape.id = id.slice(1).join('-');
+      } else if (id[0] === 'line') {
+        shape.className = 'Line';
+        var innerTags = node.querySelectorAll('*');
+        var lineTag = innerTags[0];
+        var strokeWidth = lineTag.attributes['stroke-width'].value;
+        var shift = strokeWidth % 2 !== 0 ? 0.5 : 0;
+        shape.data.x1 = lineTag.attributes.x1.value - shift;
+        shape.data.y1 = lineTag.attributes.y1.value - shift;
+        shape.data.x2 = lineTag.attributes.x2.value - shift;
+        shape.data.y2 = lineTag.attributes.y2.value - shift;
+        shape.data.strokeWidth = parseInt(strokeWidth);
+        shape.data.color = lineTag.attributes.stroke.value;
+        shape.data.capString = lineTag.attributes['stroke-linecap'].value;
+        shape.data.dash = null;
+        shape.id = id.slice(1).join('-');
+        if (typeof lineTag.attributes['stroke-dasharray'] !== 'undefined') {
+          var dash = lineTag.attributes['stroke-dasharray'].value;
+          shape.data.dash = dash.split(', ').map(a => parseInt(a));
+        }
+        shape.data.endCapShapes = [null, null];
+        if (innerTags.length > 1) {
+          for (var i = 1; i < innerTags.length; i++) {
+            var position = innerTags[i].attributes.id.value.slice(-1);
+            if (position === '0') {
```

Comment:
In general, try to reduce branching/duplication. Calculate the value of the index using the branch (varName = (position === '0') ? 0 : 1) and then set shape.data.endCapShapes[varName] = 'arrow'.

You'll need to choose a clear varName that explains what is being calculated (don't just write something like "x").

Reply:
Done.